### PR TITLE
[WIP] Ability to use an external metadata source instead of xattr

### PR DESCRIPTION
**Pull request by LcTrKiD**
Some Autopkg users are using a CI to run their recipes' list and some of them are running each recipe in an ephemeral vm with the cost of a new download every time the recipe runs. With this approach users can define a new input in the URLDownloader processor ("external_metadata_path") to change Autopkg behavior in order to use an external metadata source instead of xattr. When detect the download don't change it sets the env variable `"stop_processing_recipe"` to True.

### CODE REVIEW

**Comment by LcTrKiD** [comment]
To test this anyone can start with an empty json file `{}` and autopkg will add file metadata on its own resulting in a file with this structure: ```{"filename": {"com.github.autopkg.last-modified": "Mon, 11 Jan 2021 19:25:55 GMT", "com.github.autopkg.etag": "ffa1a6f38f7f6630015c2a998f8f76b2-16"}, "filename2": {"com.github.autopkg.last-modified": "Thu, 11 Feb 2021 23:34:40 GMT", "com.github.autopkg.etag": "9463b15ba2aae4b2213a96699e477e35-16"},...}```. A json file is a good way to store data for this but I think something external like a redis db should be more reliable.

**Comment by LcTrKiD** [comment]
The input key can't now be passed as a key/value using `-k` to `autopkg run` because it is no longer a string. The input key should be something like this:
```
<dict>
    <key>host</key>
    <string>localhost</string>
    <key>port</key>
    <integer>6379</integer>
    <key>db</key>
    <integer>0</integer>
</dict>
```
or:
```
<dict>
    <key>path</key>
    <string>test</string>
</dict>
```

File: Code/autopkglib/URLGetter.py [hunk 1]
```diff
@@ -71,6 +71,15 @@ def produce_etag_headers(self, filename):
...
+        if self.env["external_metadata_path"]:
+            metadata_object = metadata.Metadata(self.env["external_metadata_path"])
+            etag = metadata_object.getmetadata(filename, self.xattr_etag)
+            last_modified = metadata_object.getmetadata(filename, self.xattr_last_modified)
+            if etag:
+                headers["If-None-Match"] = etag
+            if last_modified:
+                headers["If-Modified-Since"] = last_modified
+            return headers
         if os.path.exists(filename):
```
Comment (homebysix) [inline Code/autopkglib/URLGetter.py:83]: I'm in favor of making this line an `elif` and keeping only a single `return headers` at the end of the function.

File: Code/autopkglib/URLDownloader.py [hunk 2]
```diff
@@ -69,6 +69,19 @@ class URLDownloader(URLGetter):
                 "'prefetch_filename' is useful for URLs with redirects."
             ),
         },
+        "external_metadata": {
```
Comment (homebysix) [inline Code/autopkglib/URLDownloader.py:72]: The URLDownloader input variable is `external_metadata`, but the URLGetter changes still reference `external_metadata_path`. Is there a discrepancy that needs fixing here?

**Review by homebysix — CHANGES_REQUESTED** [review]

**Comment by homebysix** [comment]
I tried validating this PR, but found that the provided instructions — both for providing an empty JSON file path as `external_metadata_path` and for providing an input dict as `external_metadata` — failed.

More importantly, the first thing I did — try to run `GoogleChrome.pkg` normally with no external metadata — also failed.

Please provide more detail on how this PR would be tested, and ensure that the PR doesn't break existing behavior.

File: Code/autopkglib/URLDownloader.py (hunk 2 above)
Comment (FlywireIT) [reply Code/autopkglib/URLDownloader.py:72]: That was a PEBCAK

Comment (LcTrKiD) [reply Code/autopkglib/URLDownloader.py:72]: That was a PEBCAK

File: Code/autopkglib/URLGetter.py (hunk 1 above)
Comment (LcTrKiD) [reply Code/autopkglib/URLGetter.py]: Fixed with an elif

File: Code/autopkglib/metadata.py [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+
+from redis.exceptions import ConnectionError, ResponseError
+from autopkglib import log_err
+
+
+class Metadata:
+    def __init__(self, external_metadata):
+        print(external_metadata)
+        self.metadata_object = None
+        self.host = external_metadata.get("host")
```
[the remark this answers is not in the document]
Comment (LcTrKiD) [reply Code/autopkglib/metadata.py]: Added validation in c9410d9

File: Code/autopkglib/metadata.py [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+import json
+import redis
+
+from redis.exceptions import ConnectionError, ResponseError
+from autopkglib import log_err
+
+
+class Metadata:
+    def __init__(self, external_metadata):
+        print(external_metadata)
```
[the remark this answers is not in the document]
Comment (LcTrKiD) [reply Code/autopkglib/metadata.py]: Removed line

File: Code/autopkglib/URLDownloader.py [hunk 5]
```diff
@@ -268,7 +283,11 @@ def download_changed(self, header):
 
         if header["http_result_code"] == "304":
             # resource not modified
-            self.env["download_changed"] = False
+            if self.env["external_metadata_path"]:
```
[the remark this answers is not in the document]
Comment (LcTrKiD) [reply Code/autopkglib/URLDownloader.py]: Using .get() and changed by external_metadata

File: Code/autopkglib/URLDownloader.py [hunk 6]
```diff
@@ -147,7 +159,10 @@ def prepare_download_curl_cmd(self, pathname_temporary):
         self.add_curl_common_opts(curl_cmd)
         # Clear out a potentially zero-byte file
         self.clear_zero_file(self.env["pathname"])
-        self.add_curl_headers(curl_cmd, self.produce_etag_headers(self.env["pathname"]))
+        if self.env["external_metadata_path"]:
```
[the remark this answers is not in the document]
Comment (LcTrKiD) [reply Code/autopkglib/URLDownloader.py]: Using .get() and changed by external_metadata

File: Code/autopkglib/metadata.py [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+Module that provides a way to read data from a json object
+"""
+import json
+import redis
```
[the remark this answers is not in the document]
Comment (LcTrKiD) [reply Code/autopkglib/metadata.py]: I don't like the solution proposed here c9410d9 but I'm open to ideas

**Comment by homebysix** [comment]
External metadata is a great idea and would be very useful for the growing number of AutoPkg users with CI pipelines. This particular implementation isn't workable so I'm going to close this PR, but stay tuned — I think you inspired a future attempt that will be more broadly applicable.

**Pull request closed without merging**